src/cast_core/src/RollingCollections.cpp

```diff
@@ -269,50 +295,127 @@ struct OplogTailer : public RunOperation {
-                            std::chrono::duration_cast<std::chrono::microseconds>(now - started),
-                            metrics::OutcomeType::kSuccess);
+        return (_caughtUp);
+    }
+
+    // Called when we see a create of a rolling collection.
+    // rollingMillis is the creation time taken from the name of the collection.
+    // We generally want to report the lag, but there's an issue when
+    // starting up. We'll see all the rolling collections that have been
+    // ever created in the oplog, and we'll have an artificial big spike in
+    // latencies. So we determine when we're "catching up", and ignore entries
+    // until we are caught up.
+    void trackRollingCreate(uint64_t rollingMillis, LagTrack &lagTrack) {
+        auto now = metrics::clock::now();
+        auto started = metrics::clock::time_point{
+            std::chrono::duration_cast<metrics::clock::time_point::duration>(
+              std::chrono::milliseconds{rollingMillis})};
+        auto lag = std::chrono::duration_cast<std::chrono::microseconds>(now - started);
+        auto lagns = lag.count();
+
+        if (caughtUp(lagns)) {
+            // BOOST_LOG_TRIVIAL(info) << "Oplog tailer lag: " << lagns << "ns";
+            _oplogLagOperation.report(now, lag,
+              metrics::OutcomeType::kSuccess);
+            lagTrack.addLag(lagns);
+
+            // Every minute (60 rolling collection creations), display some
+            // simple lag time stats and reset the stats.
+            if (lagTrack.count == 60) {
+                BOOST_LOG_TRIVIAL(info)
+                  << "Oplog tailer lag time stats: best " << lagTrack.best
+                  << "ns, worst " << lagTrack.worst
+                  << "ns, average " << (lagTrack.total / lagTrack.count) << "ns";
+                lagTrack.clear();
+            }
+        }
+    }
+
+    void run() override {
+        // This method should only be called once, and will only exit
+        // when the oplog has gone idle.  We are using the "await"
+        // cursor type, which always waits for the next oplog entry
+        // (or until a second elapses).  When we reach a steady state,
+        // we can determine the lag time in the oplog.  This is done
+        // by watching for the creation of rolling collections in the oplog.
+        // Each such collection is named using a time stamp, and by looking
+        // at the difference of the current time and the time stamp name,
+        // we get the lag time, including the actual creation time.
+        mongocxx::options::find opts{};
+        opts.cursor_type(mongocxx::cursor::type::k_tailable_await);
+        _cursor = std::optional<mongocxx::cursor>(database["oplog.rs"].find({}, opts));
+
+        // Track the best, worst, average lag times, we'll display them
+        // periodically to the output.
+        LagTrack lagTrack;
+
+        try {
+            // If the cursor is idle for more than 10 seconds (that is 10
```

Comment:
I'm not really a fan on this, I think the oplog tailer should stay alive regardless. It should persist as long as is specified in the duration field in the yml. If we do this we may miss some critical timeout due to replication or otherwise? Possibly miss cache stuck issues?

Reply:
Point taken about cache stuck issues.  The problem is that genny will not quit while there is a live actor (at least that's what I've seen), so I'll need a timeout of some kind.  I'll look into running at least as long as the duration.

Reply:
If you set an Actor to be `Blocking: None` (and omit `Repeat` and `Duration`) the Actor will run as long as there are other Actors running in the Phase but it won't keep the Phase (or Workload) from completing on its own.

Reply:
Example/explanation in the `HelloWorld-MultiplePhases` workload:

https://github.com/mongodb/genny/blob/869dca937569cad3f04e6ee03c1992bdbd96e819/src/workloads/docs/HelloWorld-MultiplePhases.yml#L72-L79

Reply:
Ah, okay.  I resolved this a different way, simply returning to genny when the oplog cursor returns no data.  Genny will start the actor again (or not if the time has expired).